Catalogue MultiSelect

mattigruener opened the pull request:
Hey John,

I'm trying to address parts of an internal ticket and https://github.com/GafferHQ/gaffer/issues/2675 with this PR.

There was a fairly tight dependency between plug value and selection in the ImageListing before - which I have now broken because I was assuming that we don't want to change the type of the plug to a vector type. Seemed to work well in my tests. Is that what you've been imagining as a solution, too?

Using `Delete` and `Backspace` to delete images should work as well now.

When working on the multi-select stuff, I found an issue where the new `CatalogueSelect` nodes would be added internally to a node that wraps a `Catalogue`, instead of adding it to the outside where people can see them. I believe that we were aiming to explicitly allow promoting the `imageIndex` plug to parents which suggests that what was happening was a bug? Can I get your opinion on my fix? I'm walking through the outputs of `__catalogue()['out']` to find the plug on the wrapping node that's used to pass on the Catalogue's output. Does that make sense? It works for the wrapper that we're using at IE. Seemed like the cleanest solution..?

Thanks :)

Matti.

mattigruener commented:
Added some code to allow using number keys to select images. Should help with flipping back and forth between images. We already had arrow keys that work similarly if the images are next to each other, but we don't have reordering at the moment. 

Regarding reordering - I had a look at making drag'n'drop work, but couldn't quite figure out why something like

`setDragDropMode(widget.InternalMove)`

on the internal QWidget didn't do the trick for the tree in `PathListingWidget`. We might be swallowing that signal somewhere in our wrapper layer. Do you think it's worth spending more time on that, John, or should I abandon that idea and go for either arrow buttons or a key combination like `Ctrl-up` to move images in the list? I personally like shortcuts better than using the mouse - maybe having the `Ctrl-up` feature is worth implementing independently of drag'n'drop. 
Either way - I guess currently the images are ordered based on the order in which they've been added via `addChild()`. Am I right in assuming that reordering would necessarily involve a lot of `removeChild()`/`addChild()` calls and the overhead could be quite significant for a large number of images? Am I missing something? I was looking for something along the lines of `insertChild(child, index)` which would potentially help with that. Unless I got the wrong end of the stick here, it might be better to break the connection between the ordering in the list and the parent/child mechanism? We could extend `GafferImage.Catalogue.Image` to have an index member for ordering and then make the `ImagesPath` sort the images accordingly? I'm happy to play with that tomorrow, but I thought I'd ask for your opinion first.

The internal ticket currently lists 

- [ ] reordering
- [x] multi-selection
- [x] number keys for image selection 

as the top things to look into.

Thanks!

johnhaddon commented:
> `setDragDropMode(widget.InternalMove)` on the internal QWidget didn't do the trick for the tree in PathListingWidget.

Yep,that relies on the underlying `QAbstractItemModel` supporting drag and drop (some details [here](http://doc.qt.io/archives/qt-4.8/model-view-programming.html#using-drag-and-drop-with-item-views)). Our underlying `PathModel` doesn't support that, because our `Path` class is essentially read-only at this point.

We do support drag+drop in the Plugs and [Presets](https://github.com/GafferHQ/gaffer/blob/master/python/GafferUI/UIEditor.py#L1355) listings in the UIEditor though, using our own (not particularly pretty or general) approach.

> Do you think it's worth spending more time on that, John, or should I abandon that idea and go for either arrow buttons or a key combination like Ctrl-up to move images in the list? I personally like shortcuts better than using the mouse - maybe having the Ctrl-up feature is worth implementing independently of drag'n'drop.

Personally I prefer drag and drop - I find it intuitive, easy to discover, and easier to use for things like moving one image all the way to the end. Anywhere I've added menu items or buttons or shortcuts to achieve the same thing, I've seen it as a bit of a cop-out.

> Either way - I guess currently the images are ordered based on the order in which they've been added via addChild(). Am I right in assuming that reordering would necessarily involve a lot of removeChild()/addChild() calls and the overhead could be quite significant for a large number of images? Am I missing something?

Correct. Another problem with this is that deleting an image and undoing doesn't bring it back at the same index. This is a longstanding issue with the underlying `GraphComponent` base class.

In testing, I've also discovered a bug where deleting an image and undoing does even worse than this - it's losing the correct association between images and names, which I had assumed wasn't possible. This seems to happen on master too :

1. Make a Catalogue, and add Red, Blue, Green and Yellow images in that order (dragging from a Constant is the easiest way of doing this).
2. Delete Blue, Green and Yellow (one by one in that order on master, or with multi-select on your branch).
3. Undo.
4. The image with the "Blue" label is now Yellow and vice versa.

Do you see the same at your end? I think we need to get to the bottom of this one, as I suspect it might be relevant to the question at hand.

> I was looking for something along the lines of insertChild(child, index) which would potentially help with that. Unless I got the wrong end of the stick here, it might be better to break the connection between the ordering in the list and the parent/child mechanism?

It depends a bit how you look at it. The API for the Catalogue itself relies totally on the inherent ordering of GraphComponent children - it's centered around the ordering of the children of `imagesPlug()` and the use of `indexPlug()` to pick one to view. That does have the limitations described above, but building another ordering mechanism _on top_ of it provides a whole new layer to be confused by and/or break (counterargument : that's exactly what we did for PlugLayout and NoduleLayout already, using metadata).

> We could extend GafferImage.Catalogue.Image to have an index member for ordering and then make the ImagesPath sort the images accordingly?

The good thing about the index we're already using is that it's a fundamental part of the GraphComponent data structure; it's _impossible_ for two images to have the same index. Anything we build on top won't have that same fundamental property, so it'll be a more fragile API. And I would like to fix the GraphComponent delete/undo problem anyway.

On the other hand, I find myself drawn to the solution we used for PlugLayout : leave the underlying API as-is, and just use metadata to reorder the UI presentation as a separate thing on top. Maybe let's see what gets thrown up when we get to the bottom of the bug I provided a repro for above?

mattigruener commented:
Hey John :)

Thanks for your response!

> Do you see the same at your end? 

I do see the same buggy behaviour when deleting a bunch of images and undoing that operation, yes.

Daniel also points out that what I've done with the number keys ideally would work like in nuke where using them would jump between nodes (that's in part why we wrote the CatalogueSelect node). What I've added now would add confusion once we have the nuke like behaviour. 

So, in order to move forward, how about I remove the number key stuff for now so that we can get the multi-select feature merged and then we spin off another ticket regarding the bug? Or would you rather solve the bug before merging, in order to not give users even more rope to shoot themselves in the foot with when using the shiny new multi-select to delete/undo a bunch of stuff?

Maybe we can also discuss how we want the number key shortcuts to work. Do we want extra 'View' nodes associated with the viewers and have the number shortcuts update incoming connections to them?

Thanks,

Matti.

danieldresser-ie commented:
> Do we want extra 'View' nodes associated with the viewers and have the number shortcuts update incoming connections to them?

Doesn't seem necessary to create nodes.  I think we had talked about some sort of metadata that we'd tag a node with meaning that it was assigned to slot 2, and then pressing "2" should pin that node to the viewer?  The only hard part is defining what "the viewer" means, when you could have multiple viewers.

johnhaddon commented:
> So, in order to move forward, how about I remove the number key stuff for now so that we can get the multi-select feature merged and then we spin off another ticket regarding the bug? Or would you rather solve the bug before merging, in order to not give users even more rope to shoot themselves in the foot with when using the shiny new multi-select to delete/undo a bunch of stuff?

Let's remove the number key stuff, but I'd like to get to the bottom of the bug before we roll this out - it's so much easier to trigger with multi-delete. I can take another look at it if you like, but I'm focussed on getting a release out today...

>> Do we want extra 'View' nodes associated with the viewers and have the number shortcuts update incoming connections to them?
> Doesn't seem necessary to create nodes.

I agree. Although that works nicely in Nuke, it's pretty antithetical to Gaffer's separation of the UI and the underlying node graph.

> I think we had talked about some sort of metadata that we'd tag a node with meaning that it was assigned to slot 2, and then pressing "2" should pin that node to the viewer? The only hard part is defining what "the viewer" means, when you could have multiple viewers.

Perhaps the keypress should just work on the currently focussed viewer (the one the mouse is in), and do nothing if there wasn't one? For me the bigger question is how we assign the slots in a way that doesn't feel clunky. And should this feature also be useable in other editors? Let's spawn this stuff off into a separate issue/PR since it's no longer specific to the catalogue...

mattigruener commented:
Removed the number keys feature which leaves us with

- [ ] reordering
- [x] multi-selection
- [ ] number keys for image selection 

The two other ones will be addressed separately, multi-selection will need a fix I'll be looking into.

mattigruener commented:
The last commit should fix the problem we were seeing when bulk-deleting images using multi-selection. Just reversing the order in which the images are processed did the trick.

That leaves us with the bug that existed before - when the user deletes images one-by-one in the "wrong" order. I am still happy to look into it, but it might not be as urgent anymore? The current PR should actually alleviate the pain a little now, instead of making it worse. Do you still want to handle the bug in the scope of this PR, John, or should we spin off a separate ticket?

mattigruener commented:
John, out of curiosity, can you tell me why this is gonna be a bad idea?

https://github.com/mattigruener/gaffer/commit/34f01d4d5da550e168d1e07d5678364c79cc513b

I ran some of our unittests and it seems to break stuff in `ArrayPlug`, for example. I'm lacking your big picture view regarding internals like that, so I won't look into what's actually breaking for now. I thought I could rely on you to tell me if it's worth looking into this, at all. 

For the record, the commit above seems to fix the problem we were seeing regarding multi-deleting in the wrong order. Regarding what's happening, I *think* it is this:

0. My `Catalogue` has three images: `red`, `green` and `blue`

The internal mapping on `Catalogue['__mapping']` looks like this:
```
'green':IECore.IntData( 1 ),'red':IECore.IntData( 0 ),'blue':IECore.IntData( 2 )
```
1. Delete `green`, internal mapping:

```
'red':IECore.IntData( 0 ),'blue':IECore.IntData( 1 )
```
2. Undo, causing the `InternalImage` to get added to the end of the list of nodes feeding into the internal switch, and I *think* causing the plug holding the mapping to just get reset via `ValuePlug::SetValueAction` to what it was initially? Mapping then is

```
'green':IECore.IntData( 1 ),'red':IECore.IntData( 0 ),'blue':IECore.IntData( 2 )
```
but that mapping is now out of date and blue and green seem reversed to the user because the internal images are in a different order.

I hope that makes sense.

Thanks,

Matti.

johnhaddon commented:
> John, out of curiosity, can you tell me why this is gonna be a bad idea? mattigruener@34f01d4

I don't think its a bad idea in principle - would just need to dig deeper to figure out why it's causing test failures. I suppose a big sequence of deletes and undos is potentially slower in this version, because we could end up doing repeated insertions into the middle of the vector when we undo. But `ScriptNode::deleteNodes()` already makes sure to [delete from the end first](https://github.com/GafferHQ/gaffer/blob/master/src/Gaffer/ScriptNode.cpp#L601) for performance anyway, so for the most common case that wouldn't be an issue.

> For the record, the commit above seems to fix the problem we were seeing regarding multi-deleting in the wrong order. Regarding what's happening, I think it is this...

Sounds plausible - good detective work! But on second thoughts, the mapping plug only gets used when viewing through a CatalogueSelect node doesn't it?

> Do you still want to handle the bug in the scope of this PR, John, or should we spin off a separate ticket?

Let's get it fixed now, since it's a nasty one, it's fresh in your mind and it seems like we're really close.

mattigruener commented:
Hey John,

I looked at the solution I was proposing in that experimental branch again.

The unittest error I was seeing seems to have been caused by some "magic" that removed `GraphComponents` when other `GraphComponents` got deleted. In particular when a `Plug` is removed from a `Box`, the internal `BoxIn` node gets removed, but I guess undoing that does not recreate the `BoxIn` as part of replaying the action. Instead, it gets added again, once we re-add the `Plug` to the `Box`. So, instead of this sequence

```
remove plug -> remove boxin -> <UNDO> -> add boxin -> add plug
```

we get 

```
remove plug -> remove boxin -> <UNDO> -> add plug -> add boxin
```

Does that make sense or did I misunderstand what's happening?

Using `addChildInternal` works fine with that sequence of events (if we ignore that the child is added in the wrong spot), but `insertChildInternal` gets "confused". The `BoxIn` node and the `Plug` have the same parent and the index bound to the `Action` gets invalidated when sneakily removing children from that parent. 
What I've done to fix it, is to check if the index is ok or if we should `push_back` instead because the index might be out of the vector's range. 
I'm not sure how to proceed. I don't know where else we might run into a similar issue with children getting added/removed outside of the Undo stack and where I potentially mess up things with what I'm doing in `insertChildInternal`. Maybe the worst that can happen is that the child that's being inserted is at the wrong spot in the vector? When checking for the vector's size I've also added a little bit of overhead, but that's probably negligible.

Do you want to go ahead with this fix or should I look for an alternative?

> This is a longstanding issue with the underlying GraphComponent base class.

Do you know off-hand where that problem came up before so I can check if the new behavior improves things? I've added the commit to this branch to make sure the tests succeed and that seems to be the case. 

Thanks! :)

Matti.

mattigruener commented:
Added reordering of images based on metadata. It's a little fiddly to get the indices right when using multi-selection, but I think it should be somewhat readable now. Maybe we can have a quick look at it some time this week, John :)

mattigruener commented:
Squashed in one commit that was just a fix and renamed the one that contains the new undo behaviour in `GraphComponent`.

johnhaddon commented on src/Gaffer/GraphComponent.cpp:
```diff
@@ -318,14 +345,27 @@ void GraphComponent::removeChild( GraphComponentPtr child )
 
 	if( refCount() )
 	{
+		auto children = child->parent()->children();
```
I present to you some interesting results :

```
import time

import Gaffer

s = Gaffer.ScriptNode()
for i in range( 0, 10000 ) :
	n = Gaffer.Node( "n{0}".format( i ) )
	s.addChild( n )
	s.selection().add( n )

c = time.clock()
s.deleteNodes( filter = s.selection() )
print time.clock() - c

# Master
# ------
#
# 0.06	0.06	0.05
#
# catalogueMultiSelect branch
# ---------------------------
#
# 1.95	1.91	1.92
```

We use a `vector<>` for storing children because it's the most compact memory-wise and in the most time-critical cases (hash/compute) we always access by index. But for large numbers of children it does open us up to dodgy quadratic scaling in cases like the above if we don't take great care. In [`ScriptNode::deleteNodes()`](https://github.com/GafferHQ/gaffer/blob/master/src/Gaffer/ScriptNode.cpp#L601) we deal with this by deleting the backmost nodes first, so I suspect searching from the back here is probably the answer too.

There's also the case where `addChild()` transfers a child from one parent to another and gets undone, which isn't dealt with here. And having some additional test cases is critical for something this fundamental. Unless you object, I'll take a stab at all this myself, so you can concentrate on the Catalogue stuff - sound OK?

mattigruener replied:
Sounds good. Thanks for taking care of it!

johnhaddon replied:
So, it turns out that there's really not much overhead at all in the linear search to find the index. This was the actual problem :

```
auto children = child->parent()->children();
```

Even through `children()` returns `const ChildContainer &`, `auto` chooses `ChildContainer` and so takes a copy of the entire vector...

johnhaddon commented on python/GafferImageUI/CatalogueUI.py:
```diff
@@ -281,9 +282,11 @@ def _updateFromPlug( self ) :
-			self.__nameWidget.setGraphComponent( image )
+			indices = self.__indicesFromSelection()
+			if index not in indices :
+				self.__pathListing.setSelection( IECore.PathMatcher( [ "/" + image.getName() ] ) )
```
When selecting an image, the "Name" and "Description" fields are no longer updating to reflect the selection. I suspect this is because they need to be outside the conditional here.

johnhaddon commented on python/GafferImageUI/CatalogueUI.py:
```diff
@@ -371,28 +375,55 @@ def __addClicked( self, *unused ) :
 		with Gaffer.UndoScope( self.getPlug().ancestor( Gaffer.ScriptNode ) ) :
-			self.__images().removeChild( self.__images()[index] )
-			self.getPlug().setValue( max( 0, index - 1 ) )
+			for i, index in enumerate( reversed( sorted( indices ) ) ) :
```
`i` is unused - remove the `enumerate()`

mattigruener replied:
I believe this is removed in the next commit already. Apologies for the confusion - should've squashed that hunk into the previous commit.

johnhaddon commented on python/GafferImageUI/CatalogueUI.py:
```diff
@@ -371,28 +375,55 @@ def __addClicked( self, *unused ) :
+			outPlug = plugParent["out"]
+
+			# It's possible that the Catalogue was wrapped by another node
+			if self.getPlug().parent() is not self.__catalogue():
```
This does work, but the logic was a bit hard to follow. I think the parent for the CatalogueSelect is always `self.getPlug().node().parent()`, so doesn't need to be computed from the walk. And I think the walk can be achieved in reverse as a one liner : `outPlug = next( p for p in node.children( GafferImage.ImagePlug ) if catalogue.isAncestorOf( p.source() ) )`. And these are both constant so can be lifted out of the loop, giving something much more straightforward :

```
node = self.getPlug().node()
catalogue = self.__catalogue()
outPlug = next( p for p in node.children( GafferImage.ImagePlug ) if catalogue.isAncestorOf( p.source() ) )

for index in self.__indicesFromSelection() :
	image = self.__images()[index]

	extractNode = GafferImage.CatalogueSelect()
	extractNode["in"].setInput( outPlug )
	extractNode["imageName"].setValue( image.getName() )

	node.parent().addChild( extractNode )
```

mattigruener commented:
I guess this is the first time we've pushed a PR _update_ through our new security measures - seems to have worked ok?

John, I think I've addressed your feedback. I've added a somewhat contentious commit at the end, though -> a1667dd

The commit is trying to address a problem I've noticed in my testing. Let's say we have a `Catalogue` with two images, A and B. If A is selected and we initiate a drag in the area of the `PathListingWidget` below the elements in the list, A jumps below B. The code in our respective dragMove callback just starts moving the selection to where the event is happening. That's a bit weird, isn't it?

Is anything relying on that behaviour? I've noticed that the `HierarchyView` also lets you drag the selection from anywhere into the `ScriptEditor`, for example. I don't think we'd lose a lot if that were to not work anymore?

Cheerio,

Matti.

johnhaddon commented:
> I've added a somewhat contentious commit at the end, though -> a1667dd

Makes total sense.

I noticed some odd behaviour when deleting an image after dragging to rearrange the order - it wasn't selecting the previous image in the list any more. I haven't looked into it further though so for now I'm going to put it down to the same problem that #2767 is fixing. Would be good to recheck once that is merged.

I've made one small tweak, which is to remove the special drag pointer since it kindof obscures what you're doing and makes it feel like you should be dragging the image away to some other widget. That's in keeping with the other drag-to-reorder widgets in the UIEditor. Hope that makes sense, since I've merged!

Thanks,
John

mattigruener commented:
> Hope that makes sense, since I've merged!

Totally does! Thanks for merging. I'll make sure to recheck the selection issue you mentioned once that low level change is merged. Will look at that code now to add another pair of eyes.
